consensus/src/experimental/buffer_manager.rs

```diff
@@ -150,6 +156,72 @@ impl StateManager {
         Ok(())
     }
 
+    fn try_advance_executed_root(&mut self) {
+        let mut cursor = self.execution_root.clone();
+        while cursor.is_some() {
+            {
+                let buffer_item = get_elem(&cursor);
+                if !matches!(&*buffer_item, BufferItem::Executed(_)) {
+                    break;
+                }
+            }
+            cursor = get_next(&cursor);
+        }
+        self.execution_root = cursor;
+    }
+
+    fn try_advance_signing_root(&mut self) {
+        let mut cursor = self.signing_root.clone();
+        while cursor.is_some() {
+            {
+                let buffer_item = get_elem(&cursor);
+                if !matches!(
+                    &*buffer_item,
+                    BufferItem::Signed(_) | BufferItem::Aggregated(_)
+                ) {
+                    break;
+                }
+            }
+            cursor = get_next(&cursor);
+        }
+        self.signing_root = cursor;
+    }
+
+    fn try_persisting_blocks(&mut self) {
+        let mut cursor = self.buffer.head.as_ref().cloned();
+        while cursor.is_some() {
+            let buffer_item = take_elem(&cursor);
+            if let BufferItem::Aggregated(aggregated) = buffer_item {
+                let blocks: Vec<Arc<ExecutedBlock>> = aggregated
+                    .executed_blocks
+                    .into_iter()
+                    .map(Arc::new)
+                    .collect();
+                // send to persisting phase
+                block_on(self.persisting_phase_tx.send(PersistingRequest {
```

Comment:
we can use sync channels or turn this function to async

Reply:
I think this function cannot be async because the state of the buffer is inconsistent (there is a none item) with my invariant at this moment. I'll change this channel to an unbounded sync channel.